File: 013.罗马数字转整数/solution.py
class Solution:
    def romanToInt(self, s: str) -> int:
        dic ={"I":1,"V":5,"X":10,"L":50,"C":100,"D":500,"M":1000}
        sum = 0
        l=len(s)
        pre_num = dic.get(s[0])
        for index in range(1,l):
            value=s[index]
            num = dic.get(value)
            if pre_num<num:
                sum -=pre_num
            else:
                sum+=pre_num
            pre_num = num
        sum+=pre_num
        return sum    
                        
            
            


        # special = {"IV":4,"IX":9,"XL":40,"XC":90,"CD":400,"CM":900}
        # res = 0
        # for k,v in special.items():
        #     if k in s:
        #         s = s.replace(k,str(v))   
        #         res = res+v    
                
        # if len(s)>1:        
        #     count=0
        #     for item in s:
        #         v = dic.get(item,0) 
        #         if v<10:
        #             count=count+v
        #         else:
        #             res = res + v 
        #     res =res +count  
        # else:
        #     try:
        #        res= int(s)   
        #     except:
        #         res = dic.get(s)          
        # return res                               
 
if __name__ == "__main__":
    s = Solution()
    for item in ["D","MCMXCIV"]:
        print(s.romanToInt(item))       

File: 013.罗马数字转整数/solution.go
package main
import "fmt"
func romanToInt(s string) int {
    dic:=map[string]int {"I":1,"V":5,"X":10,"L":50,"C":100,"D":500,"M":1000}
    sum:=0
    pre_num:=dic[string(s[0])]
    for i:=1;i<len(s);i++{
        value:=string(s[i])
        fmt.Println("value",value)
        num:=dic[value]
        if pre_num <num{
            sum-=pre_num
        }else{
            sum+=pre_num
        }
        pre_num = num

        }
         sum+=pre_num
        return sum    
    }  
       

func main() {
    fmt.Println(romanToInt("MCMXCIV"))
}

// class Solution:
//     def romanToInt(self, s: str) -> int:
//         dic ={"I":1,"V":5,"X":10,"L":50,"C":100,"D":500,"M":1000}
//         sum = 0
//         l=len(s)
//         pre_num = dic.get(s[0])
//         for index in range(1,l):
//             value=s[index]
//             num = dic.get(value)
//             if pre_num<num:
//                 sum -=pre_num
//             else:
//                 sum+=pre_num
//             pre_num = num
//         sum+=pre_num
//         return sum    
// 